File: src/filters/matchers.ts
/**
 * Update and message filtering system with logical combinators.
 *
 * @packageDocumentation
 */

import type { Update } from "../kernel/update.js";
import type { Message } from "../client/types.js";

export abstract class BaseFilter {
  abstract checkUpdate(update: Update): boolean | Promise<boolean>;

  and(other: BaseFilter): BaseFilter {
    return new AndFilter(this, other);
  }

  or(other: BaseFilter): BaseFilter {
    return new OrFilter(this, other);
  }

  not(): BaseFilter {
    return new NotFilter(this);
  }
}

class AndFilter extends BaseFilter {
  constructor(private f1: BaseFilter, private f2: BaseFilter) {
    super();
  }

  async checkUpdate(update: Update): Promise<boolean> {
    const res1 = await this.f1.checkUpdate(update);
    if (!res1) return false;
    return Boolean(await this.f2.checkUpdate(update));
  }
}

class OrFilter extends BaseFilter {
  constructor(private f1: BaseFilter, private f2: BaseFilter) {
    super();
  }

  async checkUpdate(update: Update): Promise<boolean> {
    const res1 = await this.f1.checkUpdate(update);
    if (res1) return true;
    return Boolean(await this.f2.checkUpdate(update));
  }
}

class NotFilter extends BaseFilter {
  constructor(private f: BaseFilter) {
    super();
  }

  async checkUpdate(update: Update): Promise<boolean> {
    const res = await this.f.checkUpdate(update);
    return !res;
  }
}

export class MessageFilter extends BaseFilter {
  constructor(private predicate: (message: Message) => boolean | Promise<boolean>) {
    super();
  }

  async checkUpdate(update: Update): Promise<boolean> {
    const message = update.effective_message;
    if (!message) return false;
    return Boolean(await this.predicate(message));
  }
}

function createMediaFilter(key: keyof Message): BaseFilter {
  return new MessageFilter((msg) => Boolean(msg[key]));
}

function createChatTypeFilter(type: string): BaseFilter {
  return new MessageFilter((msg) => msg.chat.type === type);
}

function createEntityFilter(entityType: string): BaseFilter {
  return new MessageFilter((msg) => {
    if (!msg.entities || msg.entities.length === 0) return false;
    return msg.entities.some((e) => e.type === entityType);
  });
}

const TEXT = new MessageFilter((msg) => {
  if (typeof msg.text !== "string") return false;
  if (msg.entities && msg.entities.length > 0) {
    const first = msg.entities[0];
    if (first && first.type === "bot_command" && first.offset === 0) {
      return false;
    }
  }
  return true;
});

const COMMAND = new MessageFilter((msg) => {
  if (typeof msg.text !== "string") return false;
  if (msg.entities && msg.entities.length > 0) {
    const first = msg.entities[0];
    if (first && first.type === "bot_command" && first.offset === 0) {
      return true;
    }
  }
  return false;
});

const ALL = new MessageFilter(() => true);

const PHOTO = createMediaFilter("photo");
const DOCUMENT = createMediaFilter("document");
const AUDIO = createMediaFilter("audio");
const VIDEO = createMediaFilter("video");
const VOICE = createMediaFilter("voice");
const VIDEO_NOTE = createMediaFilter("video_note");
const ANIMATION = createMediaFilter("animation");
const CONTACT = createMediaFilter("contact");
const LOCATION = createMediaFilter("location");
const VENUE = createMediaFilter("venue");
const POLL = createMediaFilter("poll");
const DICE = createMediaFilter("dice");
const STICKER = createMediaFilter("sticker");
const GAME = createMediaFilter("game");
const FORWARDED = new MessageFilter((msg) => Boolean(msg.forward_origin));
const REPLY = new MessageFilter((msg) => Boolean(msg.reply_to_message));
const CAPTION = new MessageFilter((msg) => typeof msg.caption === "string");
const ATTACHMENT = PHOTO.or(DOCUMENT).or(AUDIO).or(VIDEO).or(VOICE).or(ANIMATION);

export class RegexFilter extends BaseFilter {
  constructor(public readonly pattern: RegExp) {
    super();
  }

  async checkUpdate(update: Update): Promise<boolean> {
    const msg = update.effective_message;
    if (!msg) return false;
    const text = msg.text ?? msg.caption;
    if (typeof text !== "string") return false;

    const matches = text.match(this.pattern);
    if (!matches) return false;
    return true;
  }
}

export class CustomFilter extends BaseFilter {
  constructor(public readonly filterFn: (update: Update) => boolean | Promise<boolean>) {
    super();
  }

  async checkUpdate(update: Update): Promise<boolean> {
    return Boolean(await this.filterFn(update));
  }
}

const ChatTypePrivate = createChatTypeFilter("private");
const ChatTypeGroup = createChatTypeFilter("group");
const ChatTypeSupergroup = createChatTypeFilter("supergroup");
const ChatTypeChannel = createChatTypeFilter("channel");
const ChatTypeGroups = ChatTypeGroup.or(ChatTypeSupergroup);

export const ChatTypeFilters = {
  PRIVATE: ChatTypePrivate,
  GROUP: ChatTypeGroup,
  SUPERGROUP: ChatTypeSupergroup,
  CHANNEL: ChatTypeChannel,
  GROUPS: ChatTypeGroups,
};

const NEW_CHAT_MEMBERS = new MessageFilter((msg) => Boolean(msg.new_chat_members && msg.new_chat_members.length > 0));
const LEFT_CHAT_MEMBER = new MessageFilter((msg) => Boolean(msg.left_chat_member));
const NEW_CHAT_TITLE = new MessageFilter((msg) => Boolean(msg.new_chat_title));
const NEW_CHAT_PHOTO = new MessageFilter((msg) => Boolean(msg.new_chat_photo && msg.new_chat_photo.length > 0));
const DELETE_CHAT_PHOTO = new MessageFilter((msg) => Boolean(msg.delete_chat_photo));
const GROUP_CHAT_CREATED = new MessageFilter((msg) => Boolean(msg.group_chat_created));
const SUPERGROUP_CHAT_CREATED = new MessageFilter((msg) => Boolean(msg.supergroup_chat_created));
const CHANNEL_CHAT_CREATED = new MessageFilter((msg) => Boolean(msg.channel_chat_created));
const MIGRATE_TO_CHAT_ID = new MessageFilter((msg) => Boolean(msg.migrate_to_chat_id));
const MIGRATE_FROM_CHAT_ID = new MessageFilter((msg) => Boolean(msg.migrate_from_chat_id));
const PINNED_MESSAGE = new MessageFilter((msg) => Boolean(msg.pinned_message));
const StatusUpdateAll = NEW_CHAT_MEMBERS.or(LEFT_CHAT_MEMBER).or(NEW_CHAT_TITLE).or(NEW_CHAT_PHOTO).or(DELETE_CHAT_PHOTO).or(GROUP_CHAT_CREATED).or(SUPERGROUP_CHAT_CREATED).or(CHANNEL_CHAT_CREATED).or(MIGRATE_TO_CHAT_ID).or(MIGRATE_FROM_CHAT_ID).or(PINNED_MESSAGE);

export const StatusUpdateFilters = {
  NEW_CHAT_MEMBERS,
  LEFT_CHAT_MEMBER,
  NEW_CHAT_TITLE,
  NEW_CHAT_PHOTO,
  DELETE_CHAT_PHOTO,
  GROUP_CHAT_CREATED,
  SUPERGROUP_CHAT_CREATED,
  CHANNEL_CHAT_CREATED,
  MIGRATE_TO_CHAT_ID,
  MIGRATE_FROM_CHAT_ID,
  PINNED_MESSAGE,
  ALL: StatusUpdateAll,
};

const MENTION = createEntityFilter("mention");
const HASHTAG = createEntityFilter("hashtag");
const BOT_COMMAND = createEntityFilter("bot_command");
const URL = createEntityFilter("url");
const EMAIL = createEntityFilter("email");
const PHONE_NUMBER = createEntityFilter("phone_number");
const CASHTAG = createEntityFilter("cashtag");

export const filters = {
  ALL,
  TEXT,
  COMMAND,
  PHOTO,
  DOCUMENT,
  AUDIO,
  VIDEO,
  VOICE,
  VIDEO_NOTE,
  ANIMATION,
  CONTACT,
  LOCATION,
  VENUE,
  POLL,
  DICE,
  STICKER,
  GAME,
  FORWARDED,
  REPLY,
  CAPTION,
  ATTACHMENT,
  MENTION,
  HASHTAG,
  BOT_COMMAND,
  URL,
  EMAIL,
  PHONE_NUMBER,
  CASHTAG,
  ChatType: ChatTypeFilters,
  StatusUpdate: StatusUpdateFilters,
  Regex(pattern: RegExp | string): RegexFilter {
    return new RegexFilter(typeof pattern === "string" ? new RegExp(pattern) : pattern);
  },
  Custom(fn: (update: Update) => boolean | Promise<boolean>): CustomFilter {
    return new CustomFilter(fn);
  },
};

File: packages/go/pkg/filters/matchers.go
package filters

import (
	"github.com/Nam088/telebot/packages/go/pkg/types"
)

// Photo matches messages carrying a photo.
func Photo(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && len(msg.Photo) > 0
}

// Video matches messages carrying a video file.
func Video(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Video != nil
}

// Audio matches messages carrying an audio file.
func Audio(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Audio != nil
}

// Document matches messages carrying a general file.
func Document(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Document != nil
}

// Voice matches messages carrying a voice note.
func Voice(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Voice != nil
}

// VideoNote matches messages carrying a round video note.
func VideoNote(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.VideoNote != nil
}

// Animation matches messages carrying an animation (GIF or H.264 video).
func Animation(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Animation != nil
}

// Sticker matches messages carrying a sticker.
func Sticker(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Sticker != nil
}

// Contact matches messages carrying a shared phone contact.
func Contact(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Contact != nil
}

// Location matches messages carrying a shared location.
func Location(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Location != nil
}

// Venue matches messages carrying a venue.
func Venue(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Venue != nil
}

// Poll matches messages carrying a native poll.
func Poll(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Poll != nil
}

// Dice matches messages carrying a dice roll.
func Dice(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Dice != nil
}

// Game matches messages carrying a game.
func Game(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Game != nil
}

// Invoice matches messages carrying a payment invoice.
func Invoice(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Invoice != nil
}

// SuccessfulPayment matches service messages about a successful payment.
func SuccessfulPayment(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.SuccessfulPayment != nil
}

// Forwarded matches messages that were forwarded into the current chat,
// i.e. messages carrying a non-nil forward origin.
func Forwarded(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.ForwardOrigin != nil
}

// Reply matches messages that are replies to another message.
func Reply(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.ReplyToMessage != nil
}

// Caption matches messages with a non-empty caption.
func Caption(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && msg.Caption != ""
}

// Attachment matches messages carrying any common media attachment: a photo,
// document, audio file, video, voice note or animation.
var Attachment = Or(Photo, Document, Audio, Video, Voice, Animation)

// hasEntity reports whether msg carries a text entity of the given type.
// Only text entities are inspected, not caption entities, matching the Node
// implementation.
func hasEntity(msg *types.Message, entityType string) bool {
	for _, e := range msg.Entities {
		if e.Type == entityType {
			return true
		}
	}
	return false
}

// URL matches messages containing a "url" entity.
func URL(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && hasEntity(msg, "url")
}

// Email matches messages containing an "email" entity.
func Email(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && hasEntity(msg, "email")
}

// PhoneNumber matches messages containing a "phone_number" entity.
func PhoneNumber(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && hasEntity(msg, "phone_number")
}

// Hashtag matches messages containing a "hashtag" entity.
func Hashtag(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && hasEntity(msg, "hashtag")
}

// Cashtag matches messages containing a "cashtag" entity.
func Cashtag(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && hasEntity(msg, "cashtag")
}

// Mention matches messages containing a "mention" entity.
func Mention(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && hasEntity(msg, "mention")
}

// BotCommand matches messages containing a "bot_command" entity.
func BotCommand(u *types.Update) bool {
	msg := u.EffectiveMessage()
	return msg != nil && hasEntity(msg, "bot_command")
}

// chatTypeIs builds a Predicate that matches messages sent in a chat of the
// given Telegram chat type.
func chatTypeIs(chatType string) Predicate {
	return func(u *types.Update) bool {
		msg := u.EffectiveMessage()
		return msg != nil && msg.Chat != nil && msg.Chat.Type == chatType
	}
}

// ChatTypeFilters is the set of filters that match messages by Telegram chat
// type.
type ChatTypeFilters struct {
	// Private matches messages sent in private (one-on-one) chats.
	Private Predicate
	// Group matches messages sent in basic groups.
	Group Predicate
	// Supergroup matches messages sent in supergroups.
	Supergroup Predicate
	// Channel matches messages sent in broadcast channels.
	Channel Predicate
	// Groups matches messages sent in either basic groups or supergroups.
	Groups Predicate
}

// ChatType groups the chat-type filters: Private, Group, Supergroup, Channel
// and the combined Groups filter.
var ChatType = ChatTypeFilters{
	Private:    chatTypeIs("private"),
	Group:      chatTypeIs("group"),
	Supergroup: chatTypeIs("supergroup"),
	Channel:    chatTypeIs("channel"),
	Groups:     Or(chatTypeIs("group"), chatTypeIs("supergroup")),
}

// StatusUpdateFilters is the set of filters that match chat service messages
// such as member joins, title changes and chat migrations.
type StatusUpdateFilters struct {
	// NewChatMembers matches messages where new members joined or were added.
	NewChatMembers Predicate
	// LeftChatMember matches messages where a member left or was removed.
	LeftChatMember Predicate
	// NewChatTitle matches messages where the chat title was changed.
	NewChatTitle Predicate
	// NewChatPhoto matches messages where the chat photo was changed.
	NewChatPhoto Predicate
	// DeleteChatPhoto matches messages where the chat photo was deleted.
	DeleteChatPhoto Predicate
	// GroupChatCreated matches messages announcing a basic group was created.
	GroupChatCreated Predicate
	// SupergroupChatCreated matches messages announcing a supergroup was created.
	SupergroupChatCreated Predicate
	// ChannelChatCreated matches messages announcing a channel was created.
	ChannelChatCreated Predicate
	// MigrateToChatID matches migration service messages with a target chat ID.
	MigrateToChatID Predicate
	// MigrateFromChatID matches migration service messages with a source chat ID.
	MigrateFromChatID Predicate
	// PinnedMessage matches messages announcing a pinned message.
	PinnedMessage Predicate
	// All matches any chat status update.
	All Predicate
}

// StatusUpdate groups the filters for chat service (status update) messages.
var StatusUpdate = newStatusUpdateFilters()

func newStatusUpdateFilters() StatusUpdateFilters {
	s := StatusUpdateFilters{
		NewChatMembers: messagePredicate(func(msg *types.Message) bool {
			return len(msg.NewChatMembers) > 0
		}),
		LeftChatMember: messagePredicate(func(msg *types.Message) bool {
			return msg.LeftChatMember != nil
		}),
		NewChatTitle: messagePredicate(func(msg *types.Message) bool {
			return msg.NewChatTitle != ""
		}),
		NewChatPhoto: messagePredicate(func(msg *types.Message) bool {
			return len(msg.NewChatPhoto) > 0
		}),
		DeleteChatPhoto: messagePredicate(func(msg *types.Message) bool {
			return msg.DeleteChatPhoto
		}),
		GroupChatCreated: messagePredicate(func(msg *types.Message) bool {
			return msg.GroupChatCreated
		}),
		SupergroupChatCreated: messagePredicate(func(msg *types.Message) bool {
			return msg.SupergroupChatCreated
		}),
		ChannelChatCreated: messagePredicate(func(msg *types.Message) bool {
			return msg.ChannelChatCreated
		}),
		MigrateToChatID: messagePredicate(func(msg *types.Message) bool {
			return msg.MigrateToChatID != 0
		}),
		MigrateFromChatID: messagePredicate(func(msg *types.Message) bool {
			return msg.MigrateFromChatID != 0
		}),
		PinnedMessage: messagePredicate(func(msg *types.Message) bool {
			return msg.PinnedMessage != nil
		}),
	}
	s.All = Or(s.NewChatMembers, s.LeftChatMember, s.NewChatTitle, s.NewChatPhoto,
		s.DeleteChatPhoto, s.GroupChatCreated, s.SupergroupChatCreated,
		s.ChannelChatCreated, s.MigrateToChatID, s.MigrateFromChatID,
		s.PinnedMessage)
	return s
}
